### ISSUE TITLE
Crypto refactor

### ISSUE THREAD

**Issue by anonimal** [monero-project/kovri#57]
Ticketing for housekeeping. ~~@majestrate is currently~~  *was*  ~~working on this. See branch ```crypto_refactor```.~~

Our objective: pimpl all crypto implementations so we can switch libraries if needed.

**Comment by anonimal** [comment]
```
+EinMByte | anonimal: Good to see you've been making progress lately                                                                                                                                          │          
&anonimal | Thanks, EinMByte.                                                                                                                                                                                            │                     
&anonimal | More progress completed: I've completely finished the crypto refactor (#57).                                                                                                                                 │         
&anonimal | All crypto (cryptopp + supercop, AESNI not so much because it will be rewritten with #73) has been interfaced and is implemented behind the scenes.                                                          │         
&anonimal | Now I'll spend the next week seeing how I can improve what I've done because I'd like to see more code reusability.                                                                                          │         
&anonimal | I've also included minor design fixes that I missed in my rush to PR #170.                                                                                                                                   │         
&anonimal | All-in-all, a PR for #57 should (hopefully) be in no later than next weekend.
```

**Comment by anonimal** [comment]
```
&anonimal | I ended up throwing out work that would refactor all the pimpl work into a bridge design. Instead, I decided to keep everything uniformly pimpl.                                                             │         
&anonimal | In an effort to improve code re-useability and maintainability, I think we should (in the future) consider refactoring the refactor with a bridge pattern (or similar) in mind.
```
